syntax = "proto3";

package quick_verify;

enum RequestType {
    SendModel = 0;
    ReceiveModel = 1;
    ReceiveModelComplete = 2;
    Execute = 3;
    ExecuteComplete = 4;
 
    SendAndExecute = 5;

    SendIntermediateData = 6;
    IntermediateData = 7;

    SendModelMetaData = 8;
    SendModelMetaDataComplete = 9;

    ExchangeModel = 10;
    ExchangeModelComplete = 11;

    SendIntermediateMetaData = 12;
    SendIntermediateMetaDataComplete = 13;
    ReceiveIntermediateComplete = 14;

    

    Start = 15;
    TestExecute = 16;
    TestTransfer = 17;
    TestClear = 18;
}

message Request {
    RequestType type = 1;
    string payload = 2;

    SendModelProto send_model =3;
    ReceiveModelProto receive_model = 4;
    ReceiveModelCompleteProto receive_model_complete = 5;
    ExecuteProto execute = 6;
    ExecuteCompleteProto execute_complete = 7;
    SendAndExecuteProto send_and_execute = 8;
    SendIntermediateDataProto send_intermediate_data = 9;
    IntermediateDataProto intermediate_data = 10;
    SendModelMetaDataProto send_model_metadata = 11;
    SendModelMetaDataCompleteProto send_model_metadata_complete = 12;
    ExchangeModelProto exchange_model = 13;
    ExchangeModelCompleteProto exchange_model_complete = 14;
    SendIntermediateMetaDataProto send_intermediate_metadata= 15;
    SendIntermediateMetaDataCompleteProto send_intermediate_metadata_complete = 16;
    ReceiveIntermediateCompleteProto receive_intermediate_complete = 17;
}

message SendIntermediateMetaDataProto{
    int32 group_id = 1;
    int32 src_rank = 2;
    int32 execute_id = 3;
}

message SendIntermediateMetaDataCompleteProto{
    int32 group_id = 1;
    int32 dst_rank = 2;
    int32 execute_id = 3;
}

message ReceiveIntermediateCompleteProto{
    int32 execute_id = 1;
    int32 group_id = 2;
    int32 dst_node_id = 3;
}

message ExchangeModelProto{
    int32 recv_group_id = 1;
    int32 send_group_id = 2;
    int32 peer_rank = 3;
}

message ExchangeModelCompleteProto{
    int32 first_group_id = 1;
    int32 first_src_node_id = 2;
    int32 first_dst_node_id = 3;

    int32 second_group_id = 4;
    int32 second_src_node_id = 5;
    int32 second_dst_node_id = 6;
}

message SendAndExecuteProto{
    int32 group_id = 1;
    int32 node_id = 2;
    int32 execute_id = 3;
}

message SendModelMetaDataProto{
    int32 src_rank = 1;
    int32 group_id = 2;
    int32 layer_id = 3;
    int32 is_end = 4;
}

message SendModelMetaDataCompleteProto{
    int32 dst_rank = 1;
    int32 group_id = 2;
}

message SendModelProto{
    int32 group_id = 1;
    int32 node_id = 2;
}

message ReceiveModelProto{
    bytes param = 1;
    int32 group_id = 2;
    int32 node_id = 3;
}

message ReceiveModelCompleteProto{
    int32 group_id = 1;
    int32 src_node_id = 3;
    int32 dst_node_id = 4;
}

message ExecuteProto{
    int32 group_id = 1;
    bool is_bring_data = 2;
    int32 execute_id = 3;
    bytes input_data = 4;
}

message ExecuteCompleteProto{
    int32 group_id = 1;
    bytes output_data = 2;
    bool is_large = 3;
    int32 execute_id = 4;
    int32 node_id = 5;
}

message SendIntermediateDataProto{
    int32 group_id = 1;
    int32 node_id = 2;
    int32 execute_id = 3;
}

message IntermediateDataProto{
    int32 group_id = 1;
    int32 execute_id = 2;
    bytes input_data = 3;
}